File: homework/task-c24/task_c24.cc
/*******************************************************
Autors - Kristiāns Francis Cagulis, kc22015
C24. Izdēšs tos masīva elemetus, lai masīvs būtu augošs.
Programma izveidota: 15.10.2022.
*******************************************************/

#include <iostream>
using std::cin;
using std::cout;
using std::endl;

/*
int makeAscending(int *numbers, int length);
Funkcija makeAscending(numbers, length) -
izmet tikai tos skaitļus no masīva numbers ar elementu skaitu length, lai masīvs
numbers būtu augošs, un atgriež jaunās virknes garumu.
*/
int makeAscending(double *numbers, int length) {
  int i = 1;
  while (i < length) {
    if (numbers[i - 1] > numbers[i]) {
      // move array to the left by 1 value
      for (int j = i; j < length; j++) {
        numbers[j] = numbers[j + 1];
      }
      length--;
    } else {
      i++;
    }
  }
  return length;
}

int main() {
  int ask_continue = 1;
  do {
    int n;
    // get array length n
    while (1) {
      cout << "Enter the length of array: ";
      cin >> n;
      if (n > 0) {
        break;
      }
      cout << "Array length must be bigger than 0" << endl;
    }

    double *A = new double[n];
    // get array elements from user
    cout << "Enter array elements" << endl;
    for (int i = 0; i < n; i++) {
      cin >> A[i];
    }
    cout << endl;

    n = makeAscending(A, n);
    // print out array elements
    for (int i = 0; i < n; i++) {
      cout << A[i] << " ";
    }
    cout << endl;

    delete[] A;

    cout << "Continue program? [1/0] ";
    cin >> ask_continue;
  } while (ask_continue == 1);
  return 0;
}

File: homework/task-c24/task_c24.py
"""
Autors - Kristiāns Francis Cagulis, kc22015
C24. Izdēš tos masīva elemetus, lai masīvs būtu augošs.
Programma izveidota: 15.10.2022.
"""

import sys


def main() -> None:
    """Main function"""

    while True:
        numbers = get_input()
        print(f"{make_ascending(numbers)}")

        ask_continue = input("Continue program? [1/0] ")
        yes = ["y", "yes", "j", "ja", "jā", "", "1"]
        if ask_continue.lower().strip() not in yes:
            sys.exit("User closed the program")


def make_ascending(numbers: list) -> list:
    """Removes all elements that prevent the array from being ascending."""
    ordered_numbers = []
    previous_number = numbers[0]
    for number in numbers:
        if number >= previous_number:
            ordered_numbers.append(number)
            previous_number = number
    return ordered_numbers


def get_input() -> list:
    """Gets user input and returns its integer value"""
    while True:
        array = input("Enter numbers (q for exit): ").strip().split()
        if "q" in array:
            sys.exit("User closed the program")
        try:
            return list(map(int, array))
        except ValueError:
            print("All entered values must be integers\n")


if __name__ == "__main__":
    main()
